File: reference-implementations/validator-node/src/layer2.ts
// Layer 2 — cross-reference consistency. Rules 2.1 – 2.11.
//
// 2.11 (evidence tier floor under high-severity conditions) was added in the
// 1.8 schema revision and is enforced identically across all three validators.

import * as path from "node:path";
import { existsSync } from "node:fs";

import { makeFinding, type Finding } from "./findings.js";
import type { Manifest } from "./loader.js";

export function check(
  manifest: Manifest,
  opts: { repoRoot: string; siblings?: Record<string, Manifest> }
): Finding[] {
  const siblings = opts.siblings ?? {};
  const out: Finding[] = [];
  out.push(...rule2_1(manifest));
  out.push(...rule2_2(manifest, opts.repoRoot));
  out.push(...rule2_3(manifest, opts.repoRoot));
  out.push(...rule2_4(manifest, siblings));
  out.push(...rule2_5(manifest, siblings));
  out.push(...rule2_6(manifest));
  out.push(...rule2_7(manifest));
  out.push(...rule2_8(manifest));
  out.push(...rule2_9(manifest));
  out.push(...rule2_10(manifest));
  out.push(...rule2_11(manifest));
  return out;
}

function looksLikePath(v: unknown): boolean {
  if (typeof v !== "string" || v === "") return false;
  if (v.startsWith("http://") || v.startsWith("https://")) return false;
  return v.includes("/") || v.includes(".");
}

function rule2_1(manifest: Manifest): Finding[] {
  const surfaces = (manifest["surfaces_touched"] as unknown[]) || [];
  const primary = new Set<string>();
  for (const s of surfaces) {
    if (s && typeof s === "object" && !Array.isArray(s)) {
      const rec = s as Record<string, unknown>;
      if (rec["role"] === "primary" && typeof rec["surface"] === "string") {
        primary.add(rec["surface"] as string);
      }
    }
  }
  const covered = new Set<string>();
  const ev = (manifest["evidence_plan"] as unknown[]) || [];
  for (const e of ev) {
    if (e && typeof e === "object" && !Array.isArray(e)) {
      const surf = (e as Record<string, unknown>)["surface"];
      if (typeof surf === "string") covered.add(surf);
    }
  }
  const missing = [...primary].filter(p => !covered.has(p)).sort();
  return missing.map(s =>
    makeFinding(
      "evidence.primary_surface_required",
      "blocking",
      `primary surface ${JSON.stringify(s)} has no evidence_plan entry`
    )
  );
}

function rule2_2(manifest: Manifest, repoRoot: string): Finding[] {
  const out: Finding[] = [];
  const sots = (manifest["sot_map"] as unknown[]) || [];
  for (const s of sots) {
    if (!s || typeof s !== "object" || Array.isArray(s)) continue;
    const src = (s as Record<string, unknown>)["source"];
    if (!looksLikePath(src)) continue;
    const filePart = String(src).split(":", 1)[0]!;
    if (!existsSync(path.join(repoRoot, filePart))) {
      out.push(
        makeFinding(
          "sot.source_file_missing",
          "blocking",
          `${filePart} referenced in sot_map but not found`
        )
      );
    }
  }
  return out;
}

function rule2_3(manifest: Manifest, repoRoot: string): Finding[] {
  const out: Finding[] = [];
  const ev = (manifest["evidence_plan"] as unknown[]) || [];
  for (const e of ev) {
    if (!e || typeof e !== "object" || Array.isArray(e)) continue;
    const rec = e as Record<string, unknown>;
    if (rec["status"] !== "collected") continue;
    const loc = rec["artifact_location"];
    if (!loc) {
      out.push(
        makeFinding(
          "evidence.collected_requires_location",
          "blocking",
          "status=collected with empty artifact_location"
        )
      );
      continue;
    }
    if (typeof loc === "string" && looksLikePath(loc)) {
      if (!existsSync(path.join(repoRoot, loc))) {
        out.push(
          makeFinding(
            "evidence.artifact_missing",
            "advisory",
            `artifact_location ${loc} does not resolve`
          )
        );
      }
    }
  }
  return out;
}

function edgesFor(m: Manifest): Set<string> {
  let deps = m["depends_on"];
  if (typeof deps === "string") deps = [deps];
  if (!Array.isArray(deps)) return new Set();
  return new Set(deps.filter((d): d is string => typeof d === "string"));
}

function rule2_4(
  manifest: Manifest,
  siblings: Record<string, Manifest>
): Finding[] {
  const graph = new Map<string, Set<string>>();
  const selfId = manifest["change_id"];
  if (typeof selfId === "string" && selfId.length > 0) {
    graph.set(selfId, edgesFor(manifest));
  }
  for (const [cid, m] of Object.entries(siblings)) {
    graph.set(cid, edgesFor(m));
  }
  const cycle = findCycle(graph);
  if (cycle) {
    return [
      makeFinding(
        "decomposition.graph_must_be_acyclic",
        "blocking",
        "cycle: " + cycle.join(" -> ")
      ),
    ];
  }
  return [];
}

function findCycle(graph: Map<string, Set<string>>): string[] | null {
  const WHITE = 0,
    GRAY = 1,
    BLACK = 2;
  const color = new Map<string, number>();
  const parent = new Map<string, string | null>();
  for (const node of graph.keys()) {
    color.set(node, WHITE);
    parent.set(node, null);
  }
  const nodes = [...graph.keys()].sort();
  for (const start of nodes) {
    if (color.get(start) !== WHITE) continue;
    const result = dfs(graph, start, color, parent);
    if (result) return result;
  }
  return null;
}

function dfs(
  graph: Map<string, Set<string>>,
  start: string,
  color: Map<string, number>,
  parent: Map<string, string | null>
): string[] | null {
  const GRAY = 1,
    BLACK = 2;
  const stack: Array<{ node: string; it: Iterator<string> }> = [
    { node: start, it: [...(graph.get(start) ?? new Set())].sort()[Symbol.iterator]() },
  ];
  color.set(start, GRAY);
  while (stack.length > 0) {
    const frame = stack[stack.length - 1]!;
    const { value: nxt, done } = frame.it.next();
    if (done) {
      color.set(frame.node, BLACK);
      stack.pop();
      continue;
    }
    if (!graph.has(nxt)) continue;
    if (color.get(nxt) === GRAY) {
      const cyc: string[] = [nxt, frame.node];
      let cur = parent.get(frame.node) ?? null;
      while (cur !== null && cur !== nxt) {
        cyc.push(cur);
        cur = parent.get(cur) ?? null;
      }
      cyc.push(nxt);
      cyc.reverse();
      return cyc;
    }
    if (color.get(nxt) === 0) {
      color.set(nxt, GRAY);
      parent.set(nxt, frame.node);
      stack.push({
        node: nxt,
        it: [...(graph.get(nxt) ?? new Set())].sort()[Symbol.iterator](),
      });
    }
  }
  return null;
}

function rule2_5(
  manifest: Manifest,
  siblings: Record<string, Manifest>
): Finding[] {
  const out: Finding[] = [];
  const selfId = manifest["change_id"];
  if (typeof selfId !== "string") return out;
  const deps = (manifest["depends_on"] as unknown[]) || [];
  for (const depId of deps) {
    if (typeof depId !== "string" || !(depId in siblings)) continue;
    const blocks = (siblings[depId]!["blocks"] as unknown[]) || [];
    if (!blocks.includes(selfId)) {
      out.push(
        makeFinding(
          "decomposition.relation_must_be_bidirectional",
          "advisory",
          `${depId}.blocks is missing ${selfId}`
        )
      );
    }
  }
  return out;
}

function rule2_6(manifest: Manifest): Finding[] {
  const breaking = (manifest["breaking_change"] as Record<string, unknown>) || {};
  const level = breaking["level"];
  if (level === "L3" || level === "L4") {
    const hasTimeline = Boolean(breaking["deprecation_timeline"]);
    const hasDeprecation = Boolean(breaking["deprecation"]);
    if (!(hasTimeline || hasDeprecation)) {
      return [
        makeFinding(
          "breaking_change.l3_l4_requires_deprecation",
          "blocking",
          `breaking_change.level=${level} with neither deprecation_timeline nor deprecation marker`
        ),
      ];
    }
  }
  return [];
}

function rule2_7(manifest: Manifest): Finding[] {
  const rb = (manifest["rollback"] as Record<string, unknown>) || {};
  const mode = rb["overall_mode"] ?? rb["mode"];
  if (mode === 3 && !rb["compensation_plan"]) {
    return [
      makeFinding(
        "rollback.mode_3_requires_compensation",
        "blocking",
        "rollback.overall_mode=3 with no compensation_plan"
      ),
    ];
  }
  return [];
}

function rule2_8(manifest: Manifest): Finding[] {
  if (manifest["phase"] !== "deliver" && manifest["status"] !== "delivered") return [];
  const approvals = (manifest["approvals"] as unknown[]) || [];
  const human = approvals.some(
    a => a && typeof a === "object" && (a as Record<string, unknown>)["role"] === "human"
  );
  if (human) return [];
  return [
    makeFinding(
      "approval.human_required_for_delivery",
      "blocking",
      "phase=deliver but no approvals with role=human"
    ),
  ];
}

function rule2_9(manifest: Manifest): Finding[] {
  const surfaces = (manifest["surfaces_touched"] as unknown[]) || [];
  const hasExperience = surfaces.some(
    s => s && typeof s === "object" && (s as Record<string, unknown>)["surface"] === "experience"
  );
  if (hasExperience && !manifest["playtest"]) {
    return [
      makeFinding(
        "playtest.required_for_experience_surface",
        "blocking",
        "experience surface touched without playtest block"
      ),
    ];
  }
  return [];
}

function rule2_10(manifest: Manifest): Finding[] {
  if (manifest["phase"] === "observe" && !manifest["handoff_narrative"]) {
    return [
      makeFinding(
        "handoff.narrative_required_for_observe",
        "blocking",
        "phase=observe requires handoff_narrative"
      ),
    ];
  }
  return [];
}

// Rule 2.11 — evidence tier floor under high-severity conditions.
// At least one evidence_plan entry must be tier=critical when: breaking_change
// level >= L2, rollback mode 3, or a high-risk surface (compliance /
// real_world / experience) is touched with role=primary. Missing `tier` is
// treated as "standard" for backward compatibility with pre-1.8 manifests.

const HIGH_RISK_SURFACES = new Set(["compliance", "real_world", "experience"]);
const HIGH_BREAKING_LEVELS = new Set(["L2", "L3", "L4"]);

export function rule2_11(manifest: Manifest): Finding[] {
  const breaking = (manifest["breaking_change"] as Record<string, unknown>) || {};
  const rollback = (manifest["rollback"] as Record<string, unknown>) || {};
  const rollbackMode = rollback["overall_mode"] ?? rollback["mode"];

  const surfaces = (manifest["surfaces_touched"] as unknown[]) || [];
  const highRiskSurfacePrimary = surfaces.some(s => {
    if (!s || typeof s !== "object" || Array.isArray(s)) return false;
    const rec = s as Record<string, unknown>;
    return rec["role"] === "primary" && HIGH_RISK_SURFACES.has(rec["surface"] as string);
  });

  const highSeverity =
    HIGH_BREAKING_LEVELS.has(breaking["level"] as string) ||
    rollbackMode === 3 ||
    highRiskSurfacePrimary;

  if (!highSeverity) return [];

  const evidence = (manifest["evidence_plan"] as unknown[]) || [];
  const hasCritical = evidence.some(ev => {
    if (!ev || typeof ev !== "object" || Array.isArray(ev)) return false;
    return (ev as Record<string, unknown>)["tier"] === "critical";
  });
  if (hasCritical) return [];

  const triggers: string[] = [];
  if (HIGH_BREAKING_LEVELS.has(breaking["level"] as string)) {
    triggers.push(`breaking_change.level=${breaking["level"]}`);
  }
  if (rollbackMode === 3) triggers.push("rollback.overall_mode=3");
  if (highRiskSurfacePrimary) triggers.push("high-risk surface touched with role=primary");

  return [
    makeFinding(
      "evidence.critical_required_for_high_severity",
      "blocking",
      `at least one evidence_plan entry must be tier=critical when ${triggers.join(" / ")}`
    ),
  ];
}

File: reference-implementations/validator-python/src/agent_protocol_validate/layer2.py
"""Layer 2 — cross-reference consistency. Rules 2.1 – 2.12.

2.4 (decomposition acyclicity) and 2.5 (depends_on / blocks mirror) are the
two rules flagged as gaps in the POSIX reference; both are implemented here.
2.11 (evidence tier floor under high-severity conditions) was added in the
1.8 schema revision and is enforced identically across all three validators.
2.12 (manifest size within ceiling) was added in 1.26.0 to mechanically enforce
the prose ceiling at docs/change-manifest-spec.md §State-snapshot discipline
§Manifest size ceiling. Counts lines in the manifest file as the
tokenizer-agnostic proxy for the ~25,000-token AI-runtime read ceiling.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

from .findings import Finding


def check(
    manifest: dict[str, Any],
    *,
    repo_root: Path,
    siblings: dict[str, dict[str, Any]] | None = None,
    manifest_path: Path | None = None,
) -> list[Finding]:
    findings: list[Finding] = []
    siblings = siblings or {}

    findings.extend(_rule_2_1(manifest))
    findings.extend(_rule_2_2(manifest, repo_root))
    findings.extend(_rule_2_3(manifest, repo_root))
    findings.extend(_rule_2_4(manifest, siblings))
    findings.extend(_rule_2_5(manifest, siblings))
    findings.extend(_rule_2_6(manifest))
    findings.extend(_rule_2_7(manifest))
    findings.extend(_rule_2_8(manifest))
    findings.extend(_rule_2_9(manifest))
    findings.extend(_rule_2_10(manifest))
    findings.extend(_rule_2_11(manifest))
    if manifest_path is not None:
        findings.extend(_rule_2_12(manifest_path))

    return findings


def _looks_like_path(value: Any) -> bool:
    if not isinstance(value, str) or not value:
        return False
    if value.startswith(("http://", "https://")):
        return False
    return "/" in value or "." in value


def _rule_2_1(manifest: dict[str, Any]) -> list[Finding]:
    primary = {
        s.get("surface")
        for s in (manifest.get("surfaces_touched") or [])
        if isinstance(s, dict) and s.get("role") == "primary"
    }
    evidence_surfaces = {
        e.get("surface")
        for e in (manifest.get("evidence_plan") or [])
        if isinstance(e, dict)
    }
    missing = primary - evidence_surfaces - {None}
    return [
        Finding(
            rule_id="evidence.primary_surface_required",
            severity="blocking",
            detail=f"primary surface {s!r} has no evidence_plan entry",
        )
        for s in sorted(missing)
    ]


def _rule_2_2(manifest: dict[str, Any], repo_root: Path) -> list[Finding]:
    out: list[Finding] = []
    for sot in manifest.get("sot_map") or []:
        if not isinstance(sot, dict):
            continue
        src = sot.get("source")
        if not _looks_like_path(src):
            continue
        file_part = str(src).split(":", 1)[0]
        if not (repo_root / file_part).exists():
            out.append(
                Finding(
                    rule_id="sot.source_file_missing",
                    severity="blocking",
                    detail=f"{file_part} referenced in sot_map but not found",
                )
            )
    return out


def _rule_2_3(manifest: dict[str, Any], repo_root: Path) -> list[Finding]:
    out: list[Finding] = []
    for ev in manifest.get("evidence_plan") or []:
        if not isinstance(ev, dict) or ev.get("status") != "collected":
            continue
        loc = ev.get("artifact_location")
        if not loc:
            out.append(
                Finding(
                    rule_id="evidence.collected_requires_location",
                    severity="blocking",
                    detail="status=collected with empty artifact_location",
                )
            )
            continue
        if isinstance(loc, str) and _looks_like_path(loc):
            if not (repo_root / loc).exists():
                out.append(
                    Finding(
                        rule_id="evidence.artifact_missing",
                        severity="advisory",
                        detail=f"artifact_location {loc} does not resolve",
                    )
                )
    return out


def _rule_2_4(
    manifest: dict[str, Any], siblings: dict[str, dict[str, Any]]
) -> list[Finding]:
    graph: dict[str, set[str]] = {}

    def edges_for(m: dict[str, Any]) -> set[str]:
        deps = m.get("depends_on") or []
        if isinstance(deps, str):
            deps = [deps]
        return {d for d in deps if isinstance(d, str)}

    self_id = manifest.get("change_id")
    if isinstance(self_id, str) and self_id:
        graph[self_id] = edges_for(manifest)
    for cid, m in siblings.items():
        graph[cid] = edges_for(m)

    cycle = _find_cycle(graph)
    if cycle:
        return [
            Finding(
                rule_id="decomposition.graph_must_be_acyclic",
                severity="blocking",
                detail="cycle: " + " -> ".join(cycle),
            )
        ]
    return []


def _find_cycle(graph: dict[str, set[str]]) -> list[str] | None:
    WHITE, GRAY, BLACK = 0, 1, 2
    color = {node: WHITE for node in graph}
    parent: dict[str, str | None] = {node: None for node in graph}

    def dfs(start: str) -> list[str] | None:
        stack: list[tuple[str, iter]] = [(start, iter(sorted(graph.get(start, set()))))]
        color[start] = GRAY
        while stack:
            node, it = stack[-1]
            try:
                nxt = next(it)
            except StopIteration:
                color[node] = BLACK
                stack.pop()
                continue
            if nxt not in graph:
                continue
            if color[nxt] == GRAY:
                cyc = [nxt, node]
                cur = parent[node]
                while cur is not None and cur != nxt:
                    cyc.append(cur)
                    cur = parent[cur]
                cyc.append(nxt)
                cyc.reverse()
                return cyc
            if color[nxt] == WHITE:
                color[nxt] = GRAY
                parent[nxt] = node
                stack.append((nxt, iter(sorted(graph.get(nxt, set())))))
        return None

    for node in sorted(graph):
        if color[node] == WHITE:
            cycle = dfs(node)
            if cycle:
                return cycle
    return None


def _rule_2_5(
    manifest: dict[str, Any], siblings: dict[str, dict[str, Any]]
) -> list[Finding]:
    out: list[Finding] = []
    self_id = manifest.get("change_id")
    if not isinstance(self_id, str):
        return out
    for dep_id in manifest.get("depends_on") or []:
        if not isinstance(dep_id, str) or dep_id not in siblings:
            continue
        blocks = siblings[dep_id].get("blocks") or []
        if self_id not in blocks:
            out.append(
                Finding(
                    rule_id="decomposition.relation_must_be_bidirectional",
                    severity="advisory",
                    detail=f"{dep_id}.blocks is missing {self_id}",
                )
            )
    return out


def _rule_2_6(manifest: dict[str, Any]) -> list[Finding]:
    breaking = manifest.get("breaking_change") or {}
    level = breaking.get("level")
    if level in ("L3", "L4"):
        has_timeline = bool(breaking.get("deprecation_timeline"))
        has_deprecation = bool(breaking.get("deprecation"))
        if not (has_timeline or has_deprecation):
            return [
                Finding(
                    rule_id="breaking_change.l3_l4_requires_deprecation",
                    severity="blocking",
                    detail=(
                        f"breaking_change.level={level} with neither "
                        "deprecation_timeline nor deprecation marker"
                    ),
                )
            ]
    return []


def _rule_2_7(manifest: dict[str, Any]) -> list[Finding]:
    rollback = manifest.get("rollback") or {}
    mode = rollback.get("overall_mode") or rollback.get("mode")
    if mode == 3 and not rollback.get("compensation_plan"):
        return [
            Finding(
                rule_id="rollback.mode_3_requires_compensation",
                severity="blocking",
                detail="rollback.overall_mode=3 with no compensation_plan",
            )
        ]
    return []


def _rule_2_8(manifest: dict[str, Any]) -> list[Finding]:
    if manifest.get("phase") != "deliver" and manifest.get("status") != "delivered":
        return []
    approvals = [
        a
        for a in (manifest.get("approvals") or [])
        if isinstance(a, dict) and a.get("role") == "human"
    ]
    if not approvals:
        return [
            Finding(
                rule_id="approval.human_required_for_delivery",
                severity="blocking",
                detail="phase=deliver but no approvals with role=human",
            )
        ]
    return []


def _rule_2_9(manifest: dict[str, Any]) -> list[Finding]:
    has_experience = any(
        isinstance(s, dict) and s.get("surface") == "experience"
        for s in (manifest.get("surfaces_touched") or [])
    )
    if has_experience and not manifest.get("playtest"):
        return [
            Finding(
                rule_id="playtest.required_for_experience_surface",
                severity="blocking",
                detail="experience surface touched without playtest block",
            )
        ]
    return []


def _rule_2_10(manifest: dict[str, Any]) -> list[Finding]:
    if manifest.get("phase") == "observe" and not manifest.get("handoff_narrative"):
        return [
            Finding(
                rule_id="handoff.narrative_required_for_observe",
                severity="blocking",
                detail="phase=observe requires handoff_narrative",
            )
        ]
    return []


_HIGH_RISK_SURFACES = frozenset({"compliance", "real_world", "experience"})
_HIGH_BREAKING_LEVELS = frozenset({"L2", "L3", "L4"})


def _rule_2_11(manifest: dict[str, Any]) -> list[Finding]:
    """At least one tier=critical evidence is required when high-severity
    conditions apply: breaking_change >= L2, rollback mode 3, or a high-risk
    surface (compliance / real_world / experience) touched with role=primary.

    A missing `tier` field is treated as "standard" — pre-1.8 manifests remain
    valid. The rule fires only when the manifest itself declares high severity.
    """
    breaking = manifest.get("breaking_change") or {}
    rollback = manifest.get("rollback") or {}
    rollback_mode = rollback.get("overall_mode") or rollback.get("mode")

    high_risk_surface_primary = any(
        isinstance(s, dict)
        and s.get("role") == "primary"
        and s.get("surface") in _HIGH_RISK_SURFACES
        for s in (manifest.get("surfaces_touched") or [])
    )

    high_severity = (
        breaking.get("level") in _HIGH_BREAKING_LEVELS
        or rollback_mode == 3
        or high_risk_surface_primary
    )
    if not high_severity:
        return []

    has_critical = any(
        isinstance(ev, dict) and ev.get("tier") == "critical"
        for ev in (manifest.get("evidence_plan") or [])
    )
    if has_critical:
        return []

    triggers = []
    if breaking.get("level") in _HIGH_BREAKING_LEVELS:
        triggers.append(f"breaking_change.level={breaking.get('level')}")
    if rollback_mode == 3:
        triggers.append("rollback.overall_mode=3")
    if high_risk_surface_primary:
        triggers.append("high-risk surface touched with role=primary")
    return [
        Finding(
            rule_id="evidence.critical_required_for_high_severity",
            severity="blocking",
            detail=(
                "at least one evidence_plan entry must be tier=critical when "
                + " / ".join(triggers)
            ),
        )
    ]


_CEILING_BLOCKING_LINES = 2000
_CEILING_ADVISORY_LINES = 1500


def _rule_2_12(manifest_path: Path) -> list[Finding]:
    """Manifest size within ceiling. Mechanical enforcement of the prose
    ceiling at docs/change-manifest-spec.md §State-snapshot discipline
    §Manifest size ceiling. Lines are the tokenizer-agnostic proxy for the
    ~25,000-token AI-runtime read ceiling (typical 12-13 tokens per line of
    YAML manifest content).

    Severity ladder:
      > 2000 lines → blocking
      1500–2000   → advisory
      ≤ 1500      → none
    """
    try:
        line_count = sum(1 for _ in manifest_path.open("r", encoding="utf-8"))
    except OSError:
        return []

    if line_count > _CEILING_BLOCKING_LINES:
        return [
            Finding(
                rule_id="manifest.size_within_ceiling",
                severity="blocking",
                detail=(
                    f"manifest is {line_count} lines; ceiling is "
                    f"{_CEILING_BLOCKING_LINES} (≈25,000 tokens). Compact in "
                    "place (move verbose narrative into structured note "
                    "fields) or split via part_of per "
                    "docs/change-manifest-spec.md §Manifest size ceiling. "
                    "grep / offset-read workarounds are explicitly prohibited."
                ),
            )
        ]
    if line_count > _CEILING_ADVISORY_LINES:
        return [
            Finding(
                rule_id="manifest.size_within_ceiling",
                severity="advisory",
                detail=(
                    f"manifest is {line_count} lines; approaching ceiling at "
                    f"{_CEILING_BLOCKING_LINES}. Consider compacting "
                    "structured note fields or splitting via part_of before "
                    "the next session boundary."
                ),
            )
        ]
    return []
